rfcs/20200211-tf-types.md

```diff
@@ -0,0 +1,147 @@
+| **Updated**   | 2020-02-19                                           |
+
+## Objective
+
+This RFC proposes a new TensorFlow module and namespace (`tf.types`) dedicated to storing implementation-free type definitions, similar to C++ header files. This module has no other dependencies inside TensorFlow, so any other internal module can depend on it to ensure interoperability without the risk of creating circular dependencies. These definitions can also be used by external users, for example in pytype annotations.
+The RFC focuses on the Python API, however the design should be reviewed with cross-language consistency in mind.
+## Motivation
+
+**Interoperability and composability**. A set of standard types that formalize an interface and decouples it from implementation ensures composability between components, especially when multiple implementations are involved.
+
+**Supports the [acyclic dependencies principle](https://en.wikipedia.org/wiki/Acyclic_dependencies_principle)**. In many instances, circular dependencies are caused between low-level complex components that need to compose (e.g. autograph needs to recognize datasets, and datasets need to use autograph). Interface extraction is a common pattern for breaking such cycles.
+
+**Supports pytype**. A set of static types that is consistent under Python’s `isinstance`/`issubclass` is required to support [PEP-484 type annotations](https://www.python.org/dev/peps/pep-0484/) in TensorFlow. This module can serve as the basis for that.
+
+**Helps formalize requirements for new APIs**. Having a formal, implementation-independent definition for things such as tensors, variables, iterables, iterators makes it easy to document and test compatibility between APIs.
+
+## User Benefit
+
+Application developers may use these canonical definitions for pytype annotations.
+
+Library developers can more easily define their API interfaces by referring to this namespace.
+
+Developers of modules internal to TensorFlow can use this module to avoid creating circular dependencies.
+
+## Design Proposal
+
+### The `tf.types` Namespace / Module
+All the declarations exposed under the `tf.types` namespace reside in the `python/types/*.py` module. These are [abstract base classes](https://docs.python.org/3.7/library/abc.html) with a bare minimum of method definitions and minimal or no implementation, which serve to formalize and document the contract of common types such as `Tensor`, `Variable`, etc.
+
+These definitions may be used as PEP 484 type hints, although in some cases they may be type- or shape- erased (for example, `tf.types.Tensor` may not necessarily be parametrized by `dtype` or `shape`). Note however that designs which parametrize on shape do exist, see for instance [tensorflow#31579](https://github.com/tensorflow/tensorflow/issues/31579).
+
+The type definitions are consistent with `isinstance` and `issubclass`. For example, `isinstance(tf.Tensor, tf.types.Tensor) == True`.
```

Comment:
I believe this should not be a requirement as this is not a requirements for python types to support `isinstance`. It also restricts us: we can't use anything in the `typing` module and protocols. Ex:

```
from typing import List
import tensorflow as tf

TensorLike = Union[tf.Tensor, List[float]]
isinstance([3.8, 23.4], TensorLike)
```

```
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-10-0ff96ca65533> in <module>
----> 1 isinstance([3.8, 23.4], TensorLike)

~/softwares/python/anaconda/lib/python3.7/typing.py in __instancecheck__(self, obj)
    709
    710     def __instancecheck__(self, obj):
--> 711         return self.__subclasscheck__(type(obj))
    712
    713     def __subclasscheck__(self, cls):

~/softwares/python/anaconda/lib/python3.7/typing.py in __subclasscheck__(self, cls)
    717             if cls._special:
    718                 return issubclass(cls.__origin__, self.__origin__)
--> 719         raise TypeError("Subscripted generics cannot be used with"
    720                         " class and instance checks")
    721

TypeError: Subscripted generics cannot be used with class and instance checks
```

I believe this RFC should include some examples of implementation to make things clearer.

Reply:
Interesting. It seems like we need to think through the interaction between classical interfaces, protocols and how they play with injection mechanisms like `register_tensor_conversion_function`. Ideally we wouldn't need so many mechanisms for the same thing.

Regarding protocols, I suspect [@runtime_checkable](https://www.python.org/dev/peps/pep-0544/#runtime-checkable-decorator-and-narrowing-types-by-isinstance) can help. Although, it could come with a performance hit, as in the case of ABCMeta.

I'll look into elaborating this.

Reply:
Thanks! :)

Reply:
I added a section proposing the use of protocols for custom Tensor types - I think that is in line with the use of `typing` that you suggested as well.

Reply:
Thank you very much, I'll take a look whenever I can.